### DISCUSSION TITLE
How to instantiate generated builder for EntityView?

### DISCUSSION THREAD

**Discussion by david-kubecka** [Blazebit/blaze-persistence#1569] [Q&A] [answered]
According to docs, to obtain builder for a `View` one should call `EntityViewManager.createBuilder(View.class)`. That gives just a generic builder of the type `EntityViewBuilder`, though. The generated static builder with view-specific accessors is of the type `EntityViewBuilderBase`. How to get an instance of that?

**Comment by david-kubecka** [comment, marked as answer]
I see now that in the runtime `EntityViewManager.createBuilder` indeed returns a view-specific builder, e.g. `CatViewBuilder` which is an implementation of `EntityViewBuilderBase`.  There must be some forced casting, though, because `createBuilder` returns `EntityViewBuilder` which has the `build` but none of the view-specific getters/setters so obviously they can't be used in the source code.

**Reply by david-kubecka** [reply]
Ok, I found a way but not sure whether that's the proper way:
```
val catView = CatViewBuilder.Init(null).apply { name = "some name" }.build()
```
If it is then is there any way how to achieve that generically, e.g. by passing the view type to `EntityViewManager.createBuilder`?

**Reply by beikov** [reply]
This is the way to do it, though I'd suggest you to use `new CatViewBuilder.Init(Collections.emptyMap())` instead. Adding an additional constructor there is probably a good idea.

The strength of the builder is that it has special methods for every attribute `getName`/`setName`/`withName`. The methods of the generic builder only have generic `get(..)`/`set(..)`/`with(..)` methods and you always need to pass the attribute name or a static metamodel reference. The generated builders are independent of the `EntityViewManager` because their mainly intended usage area is when you don't have a database and hence no `EntityViewManager` available i.e. during unit tests.

>If it is then is there any way how to achieve that generically, e.g. by passing the view type to EntityViewManager.createBuilder?

Sure. You do it generically, like this:

```
var catView = entityViewManager.createBuilder(CatView.class).with(CatView_.name, "some name").build();
```

or stringly typed:

```
var catView = entityViewManager.createBuilder(CatView.class).with("name", "some name").build();
```

**Reply by david-kubecka** [reply]
Yes, my intention is to use the builder mainly in unit tests so will use the `CatViewBuilder.Init` way, possibly with some convenience wrapper ;-)

**Reply by beikov** [reply]
What kind of convenience wrapper would be needed here? The intent of the class is for you not to have to code anything in addition.

**Reply by david-kubecka** [reply]
It's just my minimalistic way of thinking :-) I would create an extension function so that I can do e.g.
```
CatViewBuilder.of { name = "some name" }
```
I believe this can be done in a generic way. I could paste here the code once I'm ready with that (not the highest priority right now, though).

**Reply by beikov** [reply]
That's a nice improvement. If it's possible for me to generate anything in addition that allows you to use it this way, I'm open to adding that.
